x/ccv/democracy/distribution/module.go

```diff
@@ -0,0 +1,132 @@
+package distribution
+
+import (
+	"time"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/cosmos/cosmos-sdk/types/module"
+
+	"github.com/cosmos/cosmos-sdk/telemetry"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	distr "github.com/cosmos/cosmos-sdk/x/distribution"
+	"github.com/cosmos/cosmos-sdk/x/distribution/keeper"
+	stakingkeeper "github.com/cosmos/cosmos-sdk/x/staking/keeper"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+	consumertypes "github.com/cosmos/interchain-security/x/ccv/consumer/types"
+
+	distrtypes "github.com/cosmos/cosmos-sdk/x/distribution/types"
+	abci "github.com/tendermint/tendermint/abci/types"
+)
+
+var (
+	_ module.AppModule           = AppModule{}
+	_ module.AppModuleBasic      = AppModuleBasic{}
+	_ module.AppModuleSimulation = AppModule{}
+)
+
+// AppModule embeds the Cosmos SDK's x/staking AppModuleBasic.
+type AppModuleBasic struct {
+	distr.AppModuleBasic
+}
+
+// AppModule embeds the Cosmos SDK's x/distribution AppModule
+type AppModule struct {
+	// embed the Cosmos SDK's x/distribution AppModule
+	distr.AppModule
+
+	keeper        keeper.Keeper
+	accountKeeper distrtypes.AccountKeeper
+	bankKeeper    distrtypes.BankKeeper
+	stakingKeeper stakingkeeper.Keeper
+
+	feeCollectorName string
+}
+
+// NewAppModule creates a new AppModule object using the native x/distribution module
+// AppModule constructor.
+func NewAppModule(
+	cdc codec.Codec, keeper keeper.Keeper, ak distrtypes.AccountKeeper,
+	bk distrtypes.BankKeeper, sk stakingkeeper.Keeper, feeCollectorName string,
+) AppModule {
+	distrAppMod := distr.NewAppModule(cdc, keeper, ak, bk, sk)
+	return AppModule{
+		AppModule:        distrAppMod,
+		keeper:           keeper,
+		accountKeeper:    ak,
+		bankKeeper:       bk,
+		stakingKeeper:    sk,
+		feeCollectorName: feeCollectorName,
+	}
+}
+
+// BeginBlocker mirror functionality of cosmos-sdk/distribution BeginBlocker
+// however it allocates no proposer reward
+func (am AppModule) BeginBlock(ctx sdk.Context, req abci.RequestBeginBlock) {
+	defer telemetry.ModuleMeasureSince(distrtypes.ModuleName, time.Now(), telemetry.MetricKeyBeginBlocker)
+
+	// TODO this is Tendermint-dependent
+	// ref https://github.com/cosmos/cosmos-sdk/issues/3095
+	if ctx.BlockHeight() > 1 {
+		am.AllocateTokens(ctx)
+	}
+
+	// record the proposer for when we payout on the next block
+	consAddr := sdk.ConsAddress(req.Header.ProposerAddress)
+	am.keeper.SetPreviousProposerConsAddr(ctx, consAddr)
+}
+
+// AllocateTokens handles distribution of the collected fees
+// bondedVotes is a list of (validator address, validator voted on last block flag) for all
+// validators in the bonded set.
+func (am AppModule) AllocateTokens(
+	ctx sdk.Context,
+) {
+
+	// fetch and clear the collected fees for distribution, since this is
+	// called in BeginBlock, collected fees will be from the previous block
+	// (and distributed to the previous proposer)
```

Comment:
Is this comment correct? There's no proposer any more

Reply:
updated comment is pushed